rustdoc: Refactor doctest collection and running code

camelid opened the pull request:
This code previously had a quite confusing structure, mixing the collection,
processing, and running of doctests with multiple layers of indirection. There
are also many cases where tons of parameters are passed to functions with little
typing information (e.g., booleans or strings are often used).

As a result, the source of bugs is obfuscated (e.g. #81070) and large changes
(e.g.  #123974) become unnecessarily complicated. This PR is a first step to try
to simplify the code and make it easier to follow and less bug-prone.

r? @GuillaumeGomez

camelid committed:
rustdoc: Rename `Tester` to `DoctestVisitor`

The new name more accurately captures what it is.

camelid committed:
rustdoc: Use `write_all` to ensure all content is written

camelid committed:
Start moving format-specific code into doctest submodule

camelid committed:
Move Markdown-specific doctest code into submodule

camelid committed:
Separate doctest collection from running

GuillaumeGomez commented:
Changes are pretty straight-forward, looks good to me. Please ping me once ready for a final review.

camelid committed:
rustdoc: Extract actual doctest running logic into function

camelid committed:
Merge `RustDoctest` and `MdDoctest` into one type

camelid committed:
Remove global options from `IndividualTestOptions`

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9322846070/job/25664752543) [(plain)](https://github.com/rust-lang/rust/commit/e809fba9f34a343b1060ea7a7e828ef730cc30e3/checks/25664752543/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.4s done
#16 DONE 34.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.54s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---

---- library/core/src/mem/maybe_uninit.rs - mem::maybe_uninit::MaybeUninit<T>::uninit_array (line 326) stdout ----
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "-m64" "/tmp/rustcyHxgup/symbols.o" "/tmp/rustdoctest5VvqmN/rust_out.rust_out.ce706d524986b34e-cgu.0.rcgu.o" "/tmp/rustdoctest5VvqmN/rust_out.3gfn498lvmaq6rnholys2wy8f.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/build/compiler_builtins-e7f0096d98e3c10b/out" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-96cdcd5fadc38e67.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-8d4ea272be51efcf.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-2bb5e22efc14393d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-02802e8e0427bc56.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-2b977fb4a386cc5d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-ab7e408df27bd365.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-6510890915b6b657.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-07124b249b89c936.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-4b0af6b67bbe8a93.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-2f5dc03a41314637.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-864730ad0ff3b603.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-1a66b96776c0eed8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-ea4449a287bc3c38.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-92f6d53e2e94c70f.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-279763368bc9fa45.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-ae57ed9e62407776.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-d0bf37205fb9f76a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-70719e8645e6f000.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-b3976a76d0927a8a.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/rustdoctest5VvqmN/rust_out" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: /usr/bin/ld: /tmp/rustdoctest5VvqmN/rust_out.rust_out.ce706d524986b34e-cgu.0.rcgu.o: in function `rust_out::main::_doctest_main_library_core_src_mem_maybe_uninit_rs_326_0::read':
          rust_out.ce706d524986b34e-cgu.0:(.text._RNvNvNvCshIRZxpdn8a8_8rust_out4main56__doctest_main_library_core_src_mem_maybe_uninit_rs_326_04read+0xf): undefined reference to `read_into_buffer'
          
          
  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#rustc-link-lib)
error: aborting due to 1 previous error

Couldn't compile the test.


failures:
    library/core/src/mem/maybe_uninit.rs - mem::maybe_uninit::MaybeUninit<T>::uninit_array (line 326)

test result: FAILED. 4631 passed; 1 failed; 42 ignored; 0 measured; 0 filtered out; finished in 56.23s

error: doctest failed, to rerun pass `-p core --doc`
  local time: Fri May 31 19:09:50 UTC 2024
  network time: Fri, 31 May 2024 19:09:50 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

camelid commented on src/librustdoc/doctest.rs:
```diff
@@ -414,20 +406,22 @@ fn run_test(
 
-    if no_run && !lang_string.compile_fail && rustdoc_options.should_persist_doctests {
+    if no_run && !lang_string.compile_fail && rustdoc_options.persist_doctests.is_none() {
+        // FIXME: why does this code check if it *shouldn't* persist doctests
+        //        -- shouldn't it be the negation?
```
This code was already there, but it's really weird. Changing it to `is_some` breaks a doctest in `core` (for `MaybeUninit::uninit_array`.

camelid committed:
rustdoc: Remove `DoctestVisitor::get_line`

This was used to get the line number of the first line from the current
docstring, which was then used together with an offset within the
docstring. It's simpler to just pass the offset to the visitor and have
it do the math because it's clearer and this calculation only needs to
be done in one place (the Rust doctest visitor).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9325707385/job/25673151487) [(plain)](https://github.com/rust-lang/rust/commit/cd9b3eb10503bc508347fb6512539c10436f03be/checks/25673151487/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/setup-python@v5' (SHA:82c7e631bb3cdc910f68e0081d67478d79c6982d)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#13 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#13 CACHED

#14 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv

#15 [8/8] COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
#15 CACHED

---
DirectMap4k:      190400 kB
DirectMap2M:     9246720 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/ada5e2c7b5427a591e30baeeee2698a5eb6db0bd/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-ada5e2c7b5427a591e30baeeee2698a5eb6db0bd-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 25.71s
##[endgroup]
fmt check
fmt: checked 4528 files
##[error]Diff in /checkout/src/librustdoc/html/markdown.rs at line 728:
 
 /// A newtype that represents a relative line number in Markdown.
-/// 
+///
+///
 /// In other words, this represents an offset from the first line of Markdown
 /// in a doc comment or other source. If the first Markdown line appears on line 32,
 /// and the `MdRelLine` is 3, then the absolute line for this one is 35. I.e., it's
##[error]Diff in /checkout/src/librustdoc/html/markdown.rs at line 735:
 /// a zero-based offset.
-pub(crate) struct MdRelLine{
+pub(crate) struct MdRelLine {
 }
 
 
##[error]Diff in /checkout/src/librustdoc/html/markdown.rs at line 740:
 impl MdRelLine {
     /// See struct docs.
-    pub(crate) const fn new(offset: usize) -> Self{
-        Self{offset}
+    pub(crate) const fn new(offset: usize) -> Self {
+        Self { offset }
 
     /// See struct docs.
     /// See struct docs.
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/etc/test-float-parse/src/bin/subnorm.rs" "/checkout/library/std/src/sys/sync/mutex/pthread.rs" "/checkout/src/etc/test-float-parse/src/bin/long-fractions.rs" "/checkout/src/etc/test-float-parse/src/bin/u64-pow2.rs" "/checkout/src/etc/test-float-parse/src/bin/u32-small.rs" "/checkout/src/etc/test-float-parse/src/bin/few-ones.rs" "/checkout/src/etc/test-float-parse/src/bin/huge-pow10.rs" "/checkout/src/etc/test-float-parse/src/bin/rand-f64.rs" "/checkout/src/etc/test-float-parse/src/bin/many-digits.rs" "/checkout/src/etc/test-float-parse/src/bin/tiny-pow10.rs" "/checkout/src/etc/test-float-parse/src/bin/short-decimals.rs" "/checkout/src/librustdoc/formats/renderer.rs" "/checkout/src/librustdoc/formats/mod.rs" "/checkout/src/librustdoc/formats/item_type.rs" "/checkout/src/librustdoc/formats/cache.rs" "/checkout/src/librustdoc/lib.rs" "/checkout/src/librustdoc/passes/check_doc_test_visibility.rs" "/checkout/src/librustdoc/passes/collect_intra_doc_links.rs" "/checkout/src/librustdoc/passes/strip_hidden.rs" "/checkout/src/librustdoc/passes/lint.rs" "/checkout/src/librustdoc/passes/mod.rs" "/checkout/src/librustdoc/passes/lint/check_code_block_syntax.rs" "/checkout/src/librustdoc/passes/lint/redundant_explicit_links.rs" "/checkout/src/librustdoc/passes/lint/unescaped_backticks.rs" "/checkout/src/librustdoc/passes/lint/html_tags.rs" "/checkout/src/librustdoc/passes/lint/bare_urls.rs" "/checkout/src/librustdoc/passes/strip_priv_imports.rs" "/checkout/src/librustdoc/passes/calculate_doc_coverage.rs" "/checkout/src/librustdoc/passes/strip_private.rs" "/checkout/src/librustdoc/passes/stripper.rs" "/checkout/src/librustdoc/passes/collect_trait_impls.rs" "/checkout/src/librustdoc/passes/check_custom_code_classes.rs" "/checkout/src/librustdoc/passes/propagate_doc_cfg.rs" "/checkout/src/librustdoc/passes/strip_aliased_non_local.rs" "/checkout/src/librustdoc/visit_lib.rs" "/checkout/src/librustdoc/lint.rs" "/checkout/src/librustdoc/clean/inline.rs" "/checkout/src/librustdoc/clean/mod.rs" "/checkout/src/librustdoc/clean/cfg/tests.rs" "/checkout/src/librustdoc/clean/types.rs" "/checkout/src/librustdoc/clean/types/tests.rs" "/checkout/src/librustdoc/clean/utils/tests.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/clean/render_macro_matchers.rs" "/checkout/src/librustdoc/clean/simplify.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/clean/blanket_impl.rs" "/checkout/src/librustdoc/clean/cfg.rs" "/checkout/src/librustdoc/externalfiles.rs" "/checkout/src/librustdoc/visit_ast.rs" "/checkout/src/librustdoc/visit.rs" "/checkout/src/librustdoc/markdown.rs" "/checkout/src/librustdoc/json/mod.rs" "/checkout/src/librustdoc/json/conversions.rs" "/checkout/src/librustdoc/json/import_finder.rs" "/checkout/src/librustdoc/html/static_files.rs" "/checkout/src/librustdoc/html/mod.rs" "/checkout/src/librustdoc/html/url_parts_builder.rs" "/checkout/src/librustdoc/html/markdown/tests.rs" "/checkout/src/librustdoc/html/format.rs" "/checkout/src/librustdoc/html/markdown.rs" "/checkout/src/librustdoc/html/url_parts_builder/tests.rs" "/checkout/src/librustdoc/html/toc.rs" "/checkout/library/std/src/sys/sync/mutex/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sat Jun  1 00:01:23 UTC 2024
  network time: Sat, 01 Jun 2024 00:01:23 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9326363074/job/25674795840) [(plain)](https://github.com/rust-lang/rust/commit/64c585d9a20979ae2052bbdab2d257a4407c992d/checks/25674795840/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#15 exporting to docker image format
#15 sending tarball 29.3s done
#15 DONE 34.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.52s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
failures:

---- [rustdoc] tests/rustdoc/test_option_check/bar.rs stdout ----

error: Not found doc test: "test /checkout/tests/rustdoc/test_option_check/bar.rs - foooo (line 5) ... ok" in "/checkout/tests/rustdoc/test_option_check/bar.rs":[6]
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/test_option_check/bar/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/test_option_check/bar" "--deny" "warnings" "/checkout/tests/rustdoc/test_option_check/bar.rs" "-A" "internal_features" "--test"

running 1 test
running 1 test
test /checkout/tests/rustdoc/test_option_check/bar.rs - foooo (line 5) ... ok
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s
------------------------------------------
stderr: none



---- [rustdoc] tests/rustdoc/test_option_check/test.rs stdout ----

error: Not found doc test: "test /checkout/tests/rustdoc/test_option_check/test.rs - Bar (line 14) ... ok" in "/checkout/tests/rustdoc/test_option_check/test.rs":[8, 15]
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/test_option_check/test/auxiliary" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/test_option_check/test" "--deny" "warnings" "/checkout/tests/rustdoc/test_option_check/test.rs" "-A" "internal_features" "--test"

running 3 tests
test /checkout/tests/rustdoc/test_option_check/test.rs - Bar (line 14) ... ok
test /checkout/tests/rustdoc/test_option_check/test.rs - Foo (line 7) ... ok
test /checkout/tests/rustdoc/test_option_check/test.rs - Foo (line 7) ... ok
test /checkout/tests/rustdoc/test_option_check/bar.rs - bar::foooo (line 5) ... ok
test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.21s
------------------------------------------
stderr: none

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9326993536/job/25676288215) [(plain)](https://github.com/rust-lang/rust/commit/60a49f8195d351a67b4ab79a7cbcf8cb9e805e68/checks/25676288215/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.7s done
#16 DONE 33.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.54s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0053]: method `visit_test` has an incompatible type for trait
    --> src/librustdoc/doctest.rs:1188:72
     |
1188 |     fn visit_test(&mut self, _test: String, _config: LangString, line: usize) {
     |                                                                        |
     |                                                                        |
     |                                                                        expected `MdRelLine`, found `usize`
     |                                                                        help: change the parameter type to match the trait: `MdRelLine`
note: type in trait
    --> src/librustdoc/doctest.rs:965:74
     |
     |
965  |     fn visit_test(&mut self, test: String, config: LangString, rel_line: MdRelLine);
     |                                                                          ^^^^^^^^^
     = note: expected signature `fn(&mut Vec<_>, std::string::String, LangString, MdRelLine)`
                found signature `fn(&mut Vec<_>, std::string::String, LangString, usize)`
For more information about this error, try `rustc --explain E0053`.
error: could not compile `rustdoc` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:31:12
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #124577) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

camelid committed:
Make two fields computed on-demand

camelid committed:
Remove `RunnableDoctest`

It should instead be the actual input to the running logic. Currently
it's not actually quite runnable since it's still missing some
information.

camelid committed:
Move some arguments to fields and reorganize fields

I moved some local arguments and options to either the local options
struct or, if it made sense, the global options struct.

camelid committed:
Make doctests before running them; reintroduce `RunnableDoctest`

camelid committed:
Move logic for "making" doctests to submodule

This code turns the raw code given by the user into something actually
runnable, e.g. by adding a `main` function if it doesn't already exist.

I also made a couple other items private that didn't need to be
crate-public.

camelid committed:
Parse full doctest source; extract helper for parsing code

It doesn't really make sense to skip part of the source when we're
parsing it, so parse the whole doctest. This simplifies things too.

camelid commented:
There's definitely more work to be done, but I think this is a good stopping point for this PR.

@GuillaumeGomez ready for review!

camelid commented:
Also I should mention that a lot of the changes listed are just moving code into files. I made sure to put those in their own commits to aid in code review.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9362870911/job/25772483822) [(plain)](https://github.com/rust-lang/rust/commit/dea33f471fb2f43132023ce9ab7c9cae792355c0/checks/25772483822/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 30.2s done
#16 DONE 35.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.50s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---

error[E0063]: missing fields `args_file` and `crate_name` in initializer of `GlobalTestOptions`
  --> src/librustdoc/doctest/tests.rs:57:9
   |
57 |         GlobalTestOptions { no_crate_inject: true, attrs: vec![], insert_indent_space: false };
   |         ^^^^^^^^^^^^^^^^^ missing `args_file` and `crate_name`
error[E0599]: no function or associated item named `default` found for struct `GlobalTestOptions` in the current scope
  --> src/librustdoc/doctest/tests.rs:75:35
   |
75 |     let opts = GlobalTestOptions::default();
---

error[E0063]: missing fields `args_file` and `crate_name` in initializer of `GlobalTestOptions`
   --> src/librustdoc/doctest/tests.rs:311:9
    |
311 |         GlobalTestOptions { no_crate_inject: false, attrs: vec![], insert_indent_space: true };
    |         ^^^^^^^^^^^^^^^^^ missing `args_file` and `crate_name`
error[E0063]: missing fields `args_file` and `crate_name` in initializer of `GlobalTestOptions`
   --> src/librustdoc/doctest/tests.rs:331:9
    |
    |
331 |         GlobalTestOptions { no_crate_inject: false, attrs: vec![], insert_indent_space: true };
    |         ^^^^^^^^^^^^^^^^^ missing `args_file` and `crate_name`
Some errors have detailed explanations: E0063, E0599.
For more information about an error, try `rustc --explain E0063`.
error: could not compile `rustdoc` (lib test) due to 18 previous errors
warning: build failed, waiting for other jobs to finish...
```

</details>

camelid committed:
Fix broken rustdoc unit tests

bors commented:
:umbrella: The latest upstream changes (presumably #126016) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9419867054/job/25950623588) [(plain)](https://github.com/rust-lang/rust/commit/3670ad59ade784226515327f2d776f02a6613f2c/checks/25950623588/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/setup-python@v5' (SHA:82c7e631bb3cdc910f68e0081d67478d79c6982d)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.422   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.440 Collecting boolean-py==4.0
#13 0.446   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.462 Collecting chardet==5.1.0
---
#13 3.636 Building wheels for collected packages: reuse
#13 3.637   Building wheel for reuse (pyproject.toml): started
#13 3.969   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.969   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.970   Stored in directory: /tmp/pip-ephem-wheel-cache-cs053o5l/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.972 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.994   Attempting uninstall: setuptools
#13 3.995     Found existing installation: setuptools 59.6.0
#13 3.996     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.996     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.997     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.664 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.665 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.181 Collecting virtualenv
#13 5.234   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.401      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 23.7 MB/s eta 0:00:00
#13 5.451 Collecting platformdirs<5,>=3.9.1
#13 5.458   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.493 Collecting filelock<4,>=3.12.2
#13 5.500   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.518 Collecting distlib<1,>=0.3.7
#13 5.527   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.537      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 62.0 MB/s eta 0:00:00
#13 5.623 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.784 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      176064 kB
DirectMap2M:     7163904 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/1be24d70ced0d6b8d41a48b6a28b3790f6facf4c/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-1be24d70ced0d6b8d41a48b6a28b3790f6facf4c-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 26.69s
##[endgroup]
fmt check
fmt: checked 5043 files
##[error]Diff in /checkout/src/librustdoc/doctest/make.rs at line 174:
             let mut found_extern_crate = crate_name.is_none();
             let mut found_macro = false;
-            let mut parser =
-            let mut parser =
-                match new_parser_from_source_str(&psess, filename, source.clone()) {
-                    Ok(p) => p,
-                    Err(errs) => {
-                        errs.into_iter().for_each(|err| err.cancel());
-                        return (found_main, found_extern_crate, found_macro);
-                };
-                };
+            let mut parser = match new_parser_from_source_str(&psess, filename, source.clone()) {
+                Ok(p) => p,
+                Err(errs) => {
+                    errs.into_iter().for_each(|err| err.cancel());
+                    return (found_main, found_extern_crate, found_macro);
+            };
 
             loop {
             loop {
                 match parser.parse_item(ForceCollect::No) {
##[error]Diff in /checkout/src/librustdoc/doctest/make.rs at line 280:
 
             let dcx = DiagCtxt::new(Box::new(emitter)).disable_warnings();
             let psess = ParseSess::with_dcx(dcx, sm);
-            let mut parser =
-                match new_parser_from_source_str(&psess, filename, source.to_owned()) {
-                    Ok(p) => p,
-                    Err(errs) => {
-                        errs.into_iter().for_each(|err| err.cancel());
-                        // tokentrees.
-                        return false;
-                    }
-                };
-                };
+            let mut parser = match new_parser_from_source_str(&psess, filename, source.to_owned()) {
+                Ok(p) => p,
+                Err(errs) => {
+                    errs.into_iter().for_each(|err| err.cancel());
+                    // tokentrees.
+                    return false;
+                }
+            };
+            };
             // If a parsing error happened, it's very likely that the attribute is incomplete.
             if let Err(e) = parser.parse_attribute(InnerAttrPolicy::Permitted) {
                 e.cancel();
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/clean/types.rs" "/checkout/src/librustdoc/clean/types/tests.rs" "/checkout/src/librustdoc/clean/utils/tests.rs" "/checkout/src/librustdoc/clean/utils.rs" "/checkout/src/librustdoc/clean/render_macro_matchers.rs" "/checkout/src/librustdoc/clean/simplify.rs" "/checkout/src/librustdoc/clean/auto_trait.rs" "/checkout/src/librustdoc/clean/blanket_impl.rs" "/checkout/src/librustdoc/clean/cfg.rs" "/checkout/src/librustdoc/externalfiles.rs" "/checkout/src/librustdoc/visit_ast.rs" "/checkout/src/librustdoc/visit.rs" "/checkout/src/librustdoc/markdown.rs" "/checkout/src/librustdoc/json/mod.rs" "/checkout/src/librustdoc/json/conversions.rs" "/checkout/src/librustdoc/json/import_finder.rs" "/checkout/src/librustdoc/html/static_files.rs" "/checkout/src/librustdoc/html/mod.rs" "/checkout/src/librustdoc/html/url_parts_builder.rs" "/checkout/src/librustdoc/html/markdown/tests.rs" "/checkout/src/librustdoc/html/format.rs" "/checkout/src/librustdoc/html/markdown.rs" "/checkout/src/librustdoc/html/url_parts_builder/tests.rs" "/checkout/src/librustdoc/html/toc.rs" "/checkout/src/librustdoc/html/escape.rs" "/checkout/src/librustdoc/html/render/type_layout.rs" "/checkout/src/librustdoc/html/render/span_map.rs" "/checkout/src/librustdoc/html/render/print_item.rs" "/checkout/src/librustdoc/html/render/write_shared.rs" "/checkout/src/librustdoc/html/render/mod.rs" "/checkout/src/librustdoc/html/render/search_index/encode.rs" "/checkout/src/librustdoc/html/render/search_index.rs" "/checkout/src/librustdoc/html/render/sidebar.rs" "/checkout/src/librustdoc/html/render/context.rs" "/checkout/src/librustdoc/html/render/tests.rs" "/checkout/src/librustdoc/html/length_limit.rs" "/checkout/src/librustdoc/html/length_limit/tests.rs" "/checkout/src/librustdoc/html/highlight.rs" "/checkout/src/librustdoc/html/highlight/tests.rs" "/checkout/src/librustdoc/html/highlight/fixtures/union.rs" "/checkout/src/librustdoc/html/highlight/fixtures/sample.rs" "/checkout/src/librustdoc/html/layout.rs" "/checkout/src/bootstrap/build.rs" "/checkout/src/bootstrap/src/lib.rs" "/checkout/src/bootstrap/src/bin/rustc.rs" "/checkout/src/bootstrap/src/bin/sccache-plus-cl.rs" "/checkout/src/bootstrap/src/bin/rustdoc.rs" "/checkout/src/bootstrap/src/bin/main.rs" "/checkout/src/librustdoc/html/tests.rs" "/checkout/src/librustdoc/html/sources.rs" "/checkout/src/librustdoc/html/toc/tests.rs" "/checkout/src/librustdoc/error.rs" "/checkout/src/librustdoc/docfs.rs" "/checkout/src/librustdoc/doctest.rs" "/checkout/src/bootstrap/src/core/config/mod.rs" "/checkout/src/bootstrap/src/core/config/flags.rs" "/checkout/src/bootstrap/src/core/config/tests.rs" "/checkout/src/bootstrap/src/core/config/config.rs" "/checkout/src/bootstrap/src/core/metadata.rs" "/checkout/src/bootstrap/src/core/builder.rs" "/checkout/src/bootstrap/src/core/mod.rs" "/checkout/src/librustdoc/doctest/markdown.rs" "/checkout/src/librustdoc/doctest/make.rs" "/checkout/src/librustdoc/clean/cfg/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Jun  7 16:08:15 UTC 2024
  network time: Fri, 07 Jun 2024 16:08:15 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GuillaumeGomez committed:
run fmt

GuillaumeGomez commented:
Thanks for the cleanup, it was very much needed! I'll give a higher priority to this PR as it is very merge conflict-prone.

@bors r+ p=1

bors commented:
:pushpin: Commit 6aab04e9b7d89af51eb96fcc35d7b3dab96f940a has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 6aab04e9b7d89af51eb96fcc35d7b3dab96f940a -->
<!-- homu: {"type":"Approved","sha":"6aab04e9b7d89af51eb96fcc35d7b3dab96f940a","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6aab04e9b7d89af51eb96fcc35d7b3dab96f940a with merge 4dc24ae394a03755cf337bebc8cd2e022a1a4848...
<!-- homu: {"type":"BuildStarted","head_sha":"6aab04e9b7d89af51eb96fcc35d7b3dab96f940a","merge_sha":"4dc24ae394a03755cf337bebc8cd2e022a1a4848"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9421226990/job/25959162297)
Approved by: GuillaumeGomez
Pushing 4dc24ae394a03755cf337bebc8cd2e022a1a4848 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"GuillaumeGomez","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9421226990/job/25959162297"},"merge_sha":"4dc24ae394a03755cf337bebc8cd2e022a1a4848"} -->

rust-timer commented:
Finished benchmarking commit (4dc24ae394a03755cf337bebc8cd2e022a1a4848): [comparison URL](https://perf.rust-lang.org/compare.html?start=e3c3ce62d7a8328120aa2abf60a34576f3862fde&end=4dc24ae394a03755cf337bebc8cd2e022a1a4848&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e3c3ce62d7a8328120aa2abf60a34576f3862fde&end=4dc24ae394a03755cf337bebc8cd2e022a1a4848&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e3c3ce62d7a8328120aa2abf60a34576f3862fde&end=4dc24ae394a03755cf337bebc8cd2e022a1a4848&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=e3c3ce62d7a8328120aa2abf60a34576f3862fde&end=4dc24ae394a03755cf337bebc8cd2e022a1a4848&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=e3c3ce62d7a8328120aa2abf60a34576f3862fde&end=4dc24ae394a03755cf337bebc8cd2e022a1a4848&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: missing data
**Artifact size**: 319.46 MiB -> 319.56 MiB (0.03%)
